### ISSUE TITLE
Refactor code into a library

### ISSUE THREAD

**Issue by mchandra** [QuazarTech/Bolt#11]
See discussion in #7

**Comment by ShyamSS-95** [comment]
I have a few doubts regarding the refactoring:
- `compute_moments(moment)` is placed under an object of class `physical_system` which is an object which is common to both the linear solver, and the non-linear solver. Right? But the computation would be dependent upon whether the moments are being calculated for the linear theory solver or the non-linear solver. How do you think we go about this? Using flags?
- What are `domain, boundaries, initial_conditions` supposed to be? Objects whose attributes hold the relevant data?
- Regarding this:`# perform a spatial FFT of (system.initial_conditions - background)`. What's the purpose for this? Do you not want a separate initialization for delta_f_hat?

**Comment by mchandra** [comment]
- Both the `linear_solver` and `nonlinear_solver` evolve the variables of the `physical_system`, such as `f`, `Ex`, etc. How they actually do it is internal to their respective classes. Given the velocity grids `v_x`, `v_y` and `v_z`, `physical_system` has all the information needed to compute moments, and this is independent of how its current state has been obtained -- either through linear or non-linear evolution.  

- I had in mind `domain`, `boundaries` and `initial_conditions` as functions.

- Right now, to set the initial state in the linear theory code, we need to linearize the initial condition in a sage worksheet and manually code it in. I was wondering if we could simply perform a spatial FFT of the initial conditions to decompose it into modes and then evolve the modes independently using the linear solver (which are all decoupled because the system has been linearized). I'll open a separate issue for this and we can put this off for later. Not crucial to the refactoring.

**Comment by ShyamSS-95** [comment]
Refer to [9d6e749](https://github.com/ShyamSS-95/Boltzmann_solver/tree/refactor_library) on branch `refactor_library` on my fork.

I've presented the intended functionality for the framework in the README.
Please also take a look at `lib/physical_system.py`. I've added the docstring to show the functionality of the `physical_system` object. This will then be passed to methods of the objects `linear_solver` and `non_linear_solver`(Haven't been implemented yet)

**Comment by mchandra** [comment]
This looks great; definitely in the right direction. I think the structure is nice and makes the code far more flexible.

1. The name `SLaPY` is terrible.
2. README looks good. Lets expand on that to include an example calculation that shows how the library should be used and display the output for a typical case, say Landau damping. We can even put a movie of phase-mixing.
2. Change `q3`, `q4`, `q5` to `p1`, `p2`, `p3` to make clear the canonical phase space coordinates `(q, p)`.
3. Change the corresponding advection terms from `T_1`, .. to `A_q1`, `A_q2`, `A_p1`, `A_p2`, `A_p3`.
4. Just like how each function in `lib/` has a docstring, it should also have a unit test to make sure the output is correct for some test input.
5. Change `non_linear_solver` to `nonlinear_solver`

**Comment by ShyamSS-95** [comment]
Please refer to [4bbaad5](https://github.com/ShyamSS-95/Boltzmann_solver/tree/refactor_library). Please check that the current approach is agreeable with you:

- Functions have been renamed in `physical_system`. Additionally, all variables have been initialized in the positionsExpanded form
- So far `nonlinear_solver` has been written to perform advection in `p1` and `p2`. The remaining functionality will be implemented once the current approach to the framework is approved.
- I've added files under `src/` which would be the format to evolve a 1D1V collisionless damping case. However, I haven't performed runs yet.

**Comment by mchandra** [comment]
Sorry, I did not notice a few critical things that I should have pointed out when you had asked for a review of 4bbaad57ba5846f32c955f5e82e54f0404bd6106 . These have also carried over to 4bbaad57ba5846f32c955f5e82e54f0404bd6106

1. `physical_system` should only contain physical details such as `initial_conditions(x, v)`, `boundary_conditions` etc. It should not contain any notion of discretization. 
2. All the details such as `N1`, `q1_center`, `da` etc are quantities needed to solve the system in a certain way and should be moved to `nonlinear_solver` and/or `linear_solver`, i.e., whereever they are needed.
3. `nonlinear_solver` will use the functions `initial_conditions(x, v)`, `A_q1(x, v)` etc defined in `physical_system` to solve the system. Something like:
```
# Inside evolve() of nonlinear_solver:
self.f_star = self.f(x - self._system.A_q1(self.p1, self.p2, self.p3) * dt)
```
4. In `lib`, make subfolders for `linear_solver` and `nonlinear_solver` and put the relevant files for each in their respective folders. For example, `interpolation_routines.py' should go into `nonlinear_solver/`
5. Do not put `log(f)` implementation inside `nonlinear_solver`. I think that should be a new user defined `physical_system` with its own collision operator implementation etc.

**Comment by ShyamSS-95** [comment]
Alright. But I still have some doubts. In PR #7, you had stated that you wanted `compute_moments` to be a method of `physical_system` where the computation of a moment of the distribution function can be performed by doing something like `system_1D_1V.compute_moments(v1**3)`. This would mean that the distribution function needs to be an attribute of the `physical_system` object. Correct? But how would that be done since the distribution function is being initialized under `linear_system` and `nonlinear_system`. Are you alright with making that `nonlinear_system.compute_moments(v1**3)` and `linear_system.compute_moments(v1**3)` instead.

**Comment by ShyamSS-95** [comment]
Additionally, it makes no sense to have `dump` in `physical_system`, since I'll need information about the DA and communicator for writing the data. Can I make `dump` a method of `linear_solver` and `nonlinear_solver`?

**Comment by mchandra** [comment]
1. Right, `linear_solver` and `nonlinear_solver` have their own distribution functions. So yes, define `compute_moments()` for them individually. Do you think there is a way for the user to inject moments into physical system? Something like:
```
system.moments = dict(density = [0, 0, 0], 
                      vx_bulk = [1, 0, 0], 
                      E = [2, 2, 2],
                      qx = [[1, 0, 0], [2, 2, 2]])
```
where `moment = [a, b, c]` are the exponents of `p1^a + p2^b + p3^c`. In the case of `qx`, these are interpreted as `(p1^1 + p2^0 + p3^0) * (p1^2 + p2^2 + p3^2)`. `linear_solver` and `nonlinear_solver` will then have functions `compute_moments()` which are used in the following manner:
```
density = nonlinear_solver.compute_moments('density')
```

2. Yes, `dump()` should also be separate for `linear_solver` and `nonlinear_solver`. When the user does `linear_system.dump()`, it dumps the solution obtained using linear evolution and likewise for the non-linear evolution. When doing `system.dump()`, it is not clear how the solution is being evolved, and as you had mentioned, it needs technical details that `physical_system` does not have (and indeed should not have).

**Comment by ShyamSS-95** [comment]
Issue was fixed in #20 #26
